File: app/src/main/java/k/bs/designpatternsp/structural/composite/ja/computer/Composite.java
package k.bs.designpatternsp.structural.composite.ja.computer;

import android.os.Build;

import androidx.annotation.RequiresApi;

import java.util.ArrayList;

public class Composite extends Equipment {

    ArrayList<Equipment> equipments = new ArrayList<>();

    Composite(String name) {
        super(0, name);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public int getPrice() {
        return equipments.stream()
                .map(Equipment::getPrice)
                .mapToInt(p -> (int) p)
                .sum();
    }

    public void add(Equipment equipment) {
        equipments.add(equipment);
    }
}

File: app/src/main/java/k/bs/designpatternsp/pattern/composite/kt/computer/Composite.kt
package k.bs.designpatternsp.pattern.composite.kt.computer

open class Composite(name: String) : Equipment(0, name) {

    private val equipments = ArrayList<Equipment>()

    override val price: Int
        get() = equipments.map { it.price }.sum()


    fun add(equipment: Equipment) =
        apply { equipments.add(equipment) }
}